src/librustc_mir/dataflow/generic.rs

```diff
@@ -0,0 +1,496 @@
+#![allow(unused)]
+
+use std::cmp::Ordering;
+use std::ops;
+
+use rustc::mir::{self, traversal, BasicBlock, Location};
+use rustc_data_structures::bit_set::BitSet;
+use rustc_data_structures::indexed_vec::{Idx, IndexVec};
+use rustc_data_structures::work_queue::WorkQueue;
+
+use crate::dataflow::BottomValue;
+
+/// A specific kind of dataflow analysis.
+///
+/// To run a dataflow analysis, one must set the initial state of the `START_BLOCK` via
+/// `initialize_start_block` and define a transfer function for each statement or terminator via
+/// the various `effect` methods. The entry set for all other basic blocks is initialized to
+/// `Self::BOTTOM_VALUE`. The dataflow `Engine` then iteratively updates the various entry sets for
+/// each block with the cumulative effects of the transfer functions of all preceding blocks.
+///
+/// You should use an `Engine` to actually run an analysis, and a `ResultsCursor` to inspect the
+/// results of that analysis like so:
+///
+/// ```ignore(cross-crate-imports)
+/// fn do_my_analysis(body: &mir::Body<'tcx>, dead_unwinds: &BitSet<BasicBlock>) {
+///     let analysis = MyAnalysis::new();
+///     let results = Engine::new(body, dead_unwinds, analysis).iterate_to_fixpoint();
+///     let mut cursor = dataflow::ResultsCursor::new(body, results);
+///
+///     for statement_index in body.block_data[START_BLOCK].statements.iter() {
+///         cursor.seek_after(Location { block: START_BLOCK, statement_index });
+///         let state = cursor.get();
+///         println!("{:?}", state);
+///     }
+/// }
+/// ```
+pub trait Analysis<'tcx>: BottomValue {
```

Comment:
Is this scheme expected to replace our existing dataflow architecture or do we need both schemes?

Reply:
Short answer: I hope to replace the current code that drives to the entry sets fixpoint with this code, but `BitDenotation` will remain (mostly) unchanged.

My long-term goal is to extend `Engine` with the graphviz debugging code, then replace the existing `DataflowBuilder` with it. Analyses that currently implement `BitDenotation` will continue to do so, and consumers of that data will have to wrap the analysis in the adapter described in [the summary](https://github.com/rust-lang/rust/pull/64566#issue-318580912) before passing it to `Engine` to be evaluated. In a perfect world, I would rename `BitDenotation` to `GenKillAnalysis`.

Before this happens, I'll also need to handle the `before_*` methods in the current `BitDenotation`. AFAICT, no analysis implements *both* the normal `effect` methods and their `before` variants, so we could do this by having every analysis implement the normal `effect` methods and have consumers decide whether those effects are observed at each statement by calling either `seek_before` or `seek_after`. However, while it is conceptually simpler, this seems more error-prone than having the analysis decide.

Reply:
> In a perfect world, I would rename BitDenotation to GenKillAnalysis.

Seems entirely reasonable to do to me. Opinions @eddyb?

This explanation is great. Can you put it somewhere on the module docs, so ppl know how to decide which analysis scheme to use?

Reply:
I wrote a summary of the current state in the latest commit with advice for people. Please take a look.

Reply:
Thanks!